src/vm.rs

```diff
@@ -549,6 +566,18 @@ where
         let (_, r1cs_var) = self.attach_variable(var);
         Expression {
             terms: vec![(r1cs_var, Scalar::one())],
+            assignment: match self.variable_assignment(var) {
+                None => None,
+                Some(v) => Some(scalar_to_u64(v)),
+            }
+        }
+    }
+
+    fn variable_assignment(&mut self, var: Variable) -> Option<Scalar> {
```

Comment:
This makes sense to return `Option<ScalarKind>` to preserve the range of the assignment.